module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -0,0 +1,579 @@
+        $auth = [
+            'tenant' => $this->config['API']['tenant'],
+            'username' => $this->config['API']['username'],
+            'password' => $this->config['API']['password'],
+        ];
+        $response = $this->makeRequest('POST', '/authn/login', json_encode($auth));
+        if ($response->getStatusCode() >= 400) {
+            throw new ILSException($response->getBody());
+        }
+        $this->token = $response->getHeaders()->get('X-Okapi-Token')
+            ->getFieldValue();
+        $this->sessionCache->folio_token = $this->token;
+    }
+
+    /**
+     * Check if our token is still valid
+     *
+     * Method taken from Stripes JS (loginServices.js:validateUser)
+     *
+     * @return void
+     */
+    protected function checkTenantToken()
+    {
+        error_log('checkTenantToken');
+        $response = $this->makeRequest('GET', '/users');
+        if ($response->getStatusCode() >= 400) {
+            $this->token = null;
+            $this->renewTenantToken();
+        }
+    }
+
+    /**
+     * Initialize the driver.
+     *
+     * Check or renew our auth token
+     *
+     * @return void
+     */
+    public function init()
+    {
+        $factory = $this->sessionFactory;
+        $this->sessionCache = $factory($this->config['API']['tenant']);
+        if ($this->sessionCache->folio_token ?? false) {
+            $this->token = $this->sessionCache->folio_token;
+        }
+        if ($this->token == null) {
+            $this->renewTenantToken();
+        } else {
+            $this->checkTenantToken();
+        }
+    }
+
+    /**
+     * Get raw object of item from inventory/items/
+     *
+     * @param string $itemId Item-level id
+     *
+     * @throw
+     * @return array
+     */
+    protected function getItem($itemId)
+    {
+        $response = $this->makeRequest('GET', '/inventory/items/' . $itemId);
+        switch ($response->getStatusCode()) {
+        case 400:
+            throw new BadRequest($response->getBody());
+        case 401:
+        case 403:
+            throw new Forbidden($response->getBody());
+        case 404:
+            throw new RecordMissing($response->getBody());
+        case 500:
+            throw new ILSException("500: Internal Server Error");
+        default:
+            return json_decode($response->getBody());
+        }
+    }
+
+    /**
+     * Get raw object of item from inventory/items/
+     *
+     * @param string $itemId Item-level id
+     *
+     * @return array
+     */
+    public function getStatus($itemId)
+    {
+        $item = $this->getItem($itemId);
+        return $item->status;
+    }
+
+    /**
+     * This method calls getStatus for an array of records or implement a bulk method
+     *
+     * @param array $idList Item-level ids
+     *
+     * @return array values from getStatus
+     */
+    public function getStatuses($idList)
+    {
+        $status = [];
+        foreach ($idList as $id) {
```

Comment:
Does FOLIO provide a mechanism for bulk-fetching statuses? That might at least be worth a TODO for the future, since iterating over multiple statuses tends to bog down performance.